packet/conn.go

```diff
@@ -24,28 +48,35 @@ type Conn struct {
 	// we removed the buffer reader because it will cause the SSLRequest to block (tls connection handshake won't be
 	// able to read the "Client Hello" data since it has been buffered into the buffer reader)
 
+	bufPool *BufPool
+
 	Sequence uint8
 }
 
 func NewConn(conn net.Conn) *Conn {
 	c := new(Conn)
 
 	c.Conn = conn
+	c.bufPool = NewBufPool()
 
 	return c
 }
 
 func (c *Conn) ReadPacket() ([]byte, error) {
-	var buf bytes.Buffer
+	// Here we use `sync.Pool` to avoid allocate/destroy buffers frequently.
+	buf := c.bufPool.Get()
+	defer c.bufPool.Return(buf)
 
-	if err := c.ReadPacketTo(&buf); err != nil {
+	if err := c.ReadPacketTo(buf); err != nil {
 		return nil, errors.Trace(err)
 	} else {
-		return buf.Bytes(), nil
+		result := make([]byte, buf.Len())
+		copy(result, buf.Bytes())
+		return result, nil
 	}
 }
 
-func (c *Conn) ReadPacketTo(w io.Writer) error {
+func (c *Conn) ReadPacketTo(w *bytes.Buffer) error {
```

Comment:
I prefer not breaking compatibility here, we can try to convert the Writer to Buffer and do the Grow below.

Reply:
Fixed